kernel/src/metaptr.rs

```diff
@@ -0,0 +1,113 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Google LLC 2024.
+
+//! Defines the MetaPtr type
+
+use core::fmt::{Formatter, LowerHex, UpperHex};
+use core::ops::AddAssign;
+
+/// A pointer with target specific metadata.
+/// This should be used any time the kernel wishes to grant authority to the user, or any time
+/// the user should be required to prove validity of a pointer.
+#[derive(Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[repr(transparent)]
+pub struct MetaPtr {
+    ptr: usize,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum MetaPermissions {
+    Any,
+    Read,
+    Write,
+    ReadWrite,
+    Execute,
+}
+
+impl From<MetaPtr> for usize {
+    #[inline]
+    fn from(from: MetaPtr) -> Self {
+        from.ptr
+    }
+}
+
+impl From<usize> for MetaPtr {
+    #[inline]
+    fn from(from: usize) -> Self {
+        Self { ptr: from }
+    }
+}
+
+impl UpperHex for MetaPtr {
+    #[inline]
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        UpperHex::fmt(&self.ptr, f)
+    }
+}
+
+impl LowerHex for MetaPtr {
+    #[inline]
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        LowerHex::fmt(&self.ptr, f)
+    }
+}
+
+impl AddAssign<usize> for MetaPtr {
```

Comment:
In a CHERI world, for example, shouldn't pointer arithmetic fail in the case that the pointer is incremented beyond the bounds? Or can this always succeed by just returning an invalid capability?

Reply:
Always succeeds, metadata may be invalid.

It was once the case that quite a few operations would trap on CHERI, but now pretty much the case that only accesses themselves do.